include/nbl/builtin/hlsl/type_traits.hlsl

```diff
@@ -425,4 +400,112 @@ struct is_matrix<matrix<T, N, M> > : bool_constant<true> {};
 }
 }
 }
+
+
+#ifdef __HLSL_VERSION
+
+
+// DXC doesn't support linking SPIR-V so this will always work I guess?
+// split because we won't be able to use `typeid` or `decltype` on functions until https://github.com/microsoft/hlsl-specs/issues/100
+#define NBL_REGISTER_TYPEID(T) namespace nbl { namespace  hlsl { namespace type_traits { namespace impl { \
+    template<> struct typeid_t<T> : integral_constant<uint32_t,__COUNTER__> {}; }}}}
```

Comment:
question, do you need to enclose in the namespaces or can you just go
```cpp
#define NBL_REGISTER_TYPEID(T) ::nbl::hlsl::type_traits::typeid_t<T> : ::nbl::hlsl::type_traits::integral_constant<uint32_t,__COUNTER__> {};
```

Reply:
likewise for `NBL_REGISTER_OBJ_TYPE`

Reply:
I've tried that but for some reason I thought it was not possible due to some other unrelated errors. We could do that  though dxc complains
```
warning: first declaration of class template specialization of 'typeid_t' outside namespace 'type_traits' is a C++11 extension [-Wc++11-extensions]
```

Reply:
if its only a warning, thats fine

Reply:
Then we should either supress the warning define a namespace wrapper macro because warnings will pollute the log

Reply:
btw its important that you "start" with `::` in front of `nbl`

Reply:
Yes we should specify to look at global namespace but I havent tested registering inside a namespace. let me test that as well

Reply:
```cpp
namespace XYZ
{
struct ABC { };
NBL_REGISTER_OBJ_TYPE(ABC) //fails
}
NBL_REGISTER_OBJ_TYPE(XYZ::ABC) // good
```

Reply:
do you know why it fails?

P.s. without "out of namespace" specializaiton you'd also have to register in the global NS